File: measuring-client/src/main/java/de/cyface/app/ui/TermsOfUseActivity.java
/*
 * Copyright 2017 Cyface GmbH
 *
 * This file is part of the Cyface App for Android.
 *
 * The Cyface App for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The Cyface App for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the Cyface App for Android. If not, see <http://www.gnu.org/licenses/>.
 */
package de.cyface.app.ui;

import static de.cyface.app.utils.Constants.ACCEPTED_TERMS_KEY;
import static de.cyface.app.utils.Constants.TAG;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;

import de.cyface.app.BuildConfig;
import de.cyface.app.R;

/**
 * The TermsOfUserActivity is the first {@link Activity} started on app launch.
 * It's responsible for informing the user about the terms of use (and data privacy conditions).
 * When the current terms are accepted or have been before, the {@link MainActivity} is launched.
 *
 * @author Armin Schnabel
 * @version 1.0.7
 * @since 1.0.0
 */
public class TermsOfUseActivity extends Activity implements View.OnClickListener {

    /**
     * Intent for switching to the main activity after this activity has been finished.
     */
    private Intent callMainActivityIntent;
    /**
     * The button to click to accept the terms of use and data privacy conditions.
     */
    private Button acceptTermsButton;
    private SharedPreferences preferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
        callMainActivityIntent = new Intent(this, MainActivity.class);

        if (currentTermsHadBeenAccepted()) {
            startActivity(callMainActivityIntent);
            finish();
        } else {
            showTermsToUser();
        }
    }

    private boolean currentTermsHadBeenAccepted() {
        int accepted_terms = preferences.getInt(ACCEPTED_TERMS_KEY, 0);
        boolean alreadyAccepted = accepted_terms == BuildConfig.currentTerms;
        if (!alreadyAccepted) {
            Log.d(TAG, "current terms (" + BuildConfig.currentTerms + ") different from accepted terms ("
                    + accepted_terms + ") -> showing terms of use.");
        }
        return alreadyAccepted;
    }

    private void showTermsToUser() {
        setContentView(R.layout.activity_terms_of_use);
        acceptTermsButton = findViewById(R.id.accept_terms_button);
        acceptTermsButton.setOnClickListener(this);
        CheckBox acceptTermsCheckbox = findViewById(R.id.accept_terms_checkbox);
        acceptTermsCheckbox
                .setOnCheckedChangeListener((buttonView, isChecked) -> acceptTermsButton.setEnabled(isChecked));
    }

    @Override
    public void onClick(View view) {
        final SharedPreferences.Editor editor = preferences.edit();
        editor.putInt(ACCEPTED_TERMS_KEY, BuildConfig.currentTerms);
        editor.apply();
        this.startActivity(callMainActivityIntent);
        finish();
    }

    @Override
    protected void onPause() {
        super.onPause();
        acceptTermsButton.setOnClickListener(null);
    }

    @Override
    public void onResume() {
        acceptTermsButton.setOnClickListener(this);
        super.onResume();
    }
}

File: ui/r4r/src/main/kotlin/de/cyface/app/r4r/TermsOfUseActivity.kt
/*
 * Copyright 2017-2025 Cyface GmbH
 *
 * This file is part of the Cyface App for Android.
 *
 * The Cyface App for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The Cyface App for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the Cyface App for Android. If not, see <http://www.gnu.org/licenses/>.
 */
package de.cyface.app.r4r

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.CheckBox
import android.widget.CompoundButton
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import de.cyface.utils.settings.AppSettings
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

/**
 * The TermsOfUserActivity is the first [Activity] started on app launch.
 *
 * It's responsible for allow the user to opt-in for Sentry.
 *
 * When the current terms are accepted or have been before, the [MainActivity] is launched.
 *
 * @author Armin Schnabel
 */
class TermsOfUseActivity : AppCompatActivity(), View.OnClickListener {
    /**
     * Intent for switching to the main activity after this activity has been finished.
     */
    private var callMainActivityIntent: Intent? = null

    /**
     * The button to click to accept the terms of use and data privacy conditions.
     */
    private var acceptTermsButton: Button? = null

    /**
     * `True` if the user opted-in to error reporting.
     */
    private var isReportingEnabled = false

    /**
     * The settings used by both, UIs and libraries.
     */
    private lateinit var appSettings: AppSettings

    /**
     * Allows the user to opt-in to error reporting.
     */
    private var acceptReportsCheckbox: CheckBox? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        appSettings = Application.appSettings
        callMainActivityIntent = Intent(this, MainActivity::class.java)
        if (currentTermsHadBeenAccepted()) {
            startActivity(callMainActivityIntent)
            finish()
            return
        }
        setContentView(R.layout.activity_terms_of_use)
    }

    /**
     * @return `True` if the latest privacy policy was accepted by the user.
     */
    private fun currentTermsHadBeenAccepted(): Boolean {
        val acceptedTerms = runBlocking { appSettings.acceptedTermsFlow.first() }
        return acceptedTerms == BuildConfig.currentTerms
    }

    /**
     * Registers the handlers for user interaction.
     */
    private fun registerOnClickListeners() {
        acceptTermsButton = findViewById(R.id.accept_terms_button)
        acceptTermsButton!!.setOnClickListener(this)
        acceptReportsCheckbox = findViewById(R.id.accept_reports_checkbox)
        acceptReportsCheckbox!!
            .setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                isReportingEnabled = isChecked
            }
    }

    /**
     * Unregisters the handlers for user interaction.
     */
    private fun unregisterOnClickListeners() {
        acceptTermsButton!!.setOnClickListener(null)
        acceptReportsCheckbox!!.setOnCheckedChangeListener(null)
    }

    override fun onClick(view: View) {
        lifecycleScope.launch {
            appSettings.setAcceptedTerms(BuildConfig.currentTerms)
            appSettings.setReportErrors(isReportingEnabled)
        }
        this.startActivity(callMainActivityIntent)
        finish()
    }

    override fun onPause() {
        super.onPause()
        unregisterOnClickListeners()
        acceptTermsButton!!.setOnClickListener(null)
    }

    public override fun onResume() {
        registerOnClickListeners()
        super.onResume()
    }
}